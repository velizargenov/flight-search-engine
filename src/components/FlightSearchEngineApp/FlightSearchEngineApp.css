h1 {
  text-decoration: underline;
}
#root {
  border: 1px solid;
  width: 1024px;
  height: 700px;
  margin: 0 auto;
  position: absolute;
  top: 50%;
  left: 50%;
  margin-top: -350px;
  margin-left: -512px;
  overflow: hidden;
}
#root > main {
  height: 582px; 
}

@media only screen and (max-width: 736px) {
  #root {
    width: 340px;
    height: 640px;
    margin-top: -320px;
    margin-left: -170px;
  }
  #root > main {
    height: 587px;
  }
}
@media only screen and (max-height: 414px) {
  #root {
    border: 1px solid black;
    width: 600px;
    height: 320px;
    margin-top: -160px;
    margin-left: -300px;
  }
}