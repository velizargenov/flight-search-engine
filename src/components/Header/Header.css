header {
  border-bottom: 1px solid #000000;
  height: 150px;
  line-height: 190px;
  padding-left: 7px;
  font-size: 52px;
  font-weight: 600;
  font-family: "Open Sans";
  letter-spacing: 1px;
}
