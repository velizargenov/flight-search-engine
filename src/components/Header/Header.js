import React from 'react';

const Header = () => (
  <h2>Header Component</h2>
);

export default Header;
