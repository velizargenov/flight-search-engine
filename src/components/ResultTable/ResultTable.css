.result-table {
  width: 746px;
  padding: 21px 10px 0 10px;
  display: inline-block;
  overflow-y: auto;
  height: 563px;
  position: relative;
}
.result-table::-webkit-scrollbar {
  display: none;
  -ms-overflow-style: -ms-autohiding-scrollbar;
}
.result-table > h2 {
  font-size: 18px;
  font-weight: 400;
  letter-spacing: 1px;
  margin: 0;
  padding: 0;
}
.result-table > section {
  display: inline-block;
  overflow: hidden;
  max-width: 200px;
  position: absolute;
  right: 10px;
  top: 0;
  top: 21px;
}
.result-table--summary {
  display: block;
  overflow: hidden;
  max-width: 200px;
  max-height: 13px;
  font-size: 11px;
  letter-spacing: 0.7px;
  margin-bottom: 3px;
}
.result-table--summary:last-child {
  margin-bottom: 0;
}
.result-table--summary > .label {
  margin-right: 5px;
}
