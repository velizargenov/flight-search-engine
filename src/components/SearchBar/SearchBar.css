.search-bar {
  border-right: 1px solid #000000;
  width: 236px;
  height: 542px;
  padding: 21px 10px;
  float: left;
}
.search-bar form {
  border: 1px solid black;
  min-height: 345px;
}
.input {
  display: inline-block;
  width: 212px; 
  padding: 13px 10px 0px 10px;
  /* width: 216px; */
}
.input input {
  border: 1px solid black;
  width: 208px;
  height: 20px;
  padding: 0 3px
}
.input--text {
  display: block;
  font-size: 12px;
  letter-spacing: 0.5px;
  margin: 0 0 2px 3px;
}
.search-bar--date {
  margin-left: 10px;
  margin-top: 9px;
  position: relative;
}
.search-bar--arrow {
  width: 27px;
  height: 22px;
  position: absolute;
  border-left: 1px solid #000000;
  border-right: 1px solid #000000;
  right: 10px;
  z-index: 999;
  top: 17px;
  background: rgb(207, 207, 207);
  cursor: pointer;
}
.search-bar--arrow:after {
  content: '';
  width: 0;
  height: 0;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  border-top: 10px solid #000000;
  position: absolute;
  top: 6px;
  left: 5px;
}
.search-bar--button-search {
  margin-left: 55px;
  margin: 35px 0 0 55px;
}