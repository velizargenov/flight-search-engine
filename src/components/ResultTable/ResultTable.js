import React from 'react';

import Flight from '../Flight/Flight';
import './ResultTable.css';

const ResultTable = ({
  flights, isReturnFlight, departureDate, returnDate,
}) => (
  <div className="result-table">
    <div className="result-table--header">
      <h2>Your Results</h2>
      <section >
        <span className="result-table--summary depart">
          <span className="label">Depart: <span>{departureDate}</span></span>
        </span>
        <span className="result-table--summary return">
          {
            isReturnFlight &&
            <span className="label">Return: <span>{returnDate}</span></span>
          }
        </span>
      </section>
    </div>
    {/* Handle logic if it return or one way to pass the correct data to <Flight />
    * implement a handleOneWayAndReturn() method
    */}
    {
      flights.map(flight => <Flight key={flight.id} details={flight} />)
    }
  </div>
);

export default ResultTable;
