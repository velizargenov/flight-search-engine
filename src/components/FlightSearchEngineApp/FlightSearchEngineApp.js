import React, { Component } from 'react';
import moment from 'moment';
import axios from 'axios';
import './FlightSearchEngineApp.css';

import flights from '../../data';
import { formatDate, isValidInput, indicateActiveButton, calculateFlightPrice } from '../helpers';
import Header from '../Header/Header';
import SearchBar from '../SearchBar/SearchBar';
import ResultTable from '../ResultTable/ResultTable';

class FlightSearchEngineApp extends Component {
  constructor (props) {
    super(props);

    this.state = {
      isReturnFlight: true,
      from: '',
      destination: '',
      departureDate: formatDate(moment('2018-01-18')),
      returnDate: 'not selected',
      numberOfPassengers: '1',
      rangeValue: {
        min: 0, max: 200,
      },
      userHasSearched: false,
      data: [],
      flights: [],
    };

    this.handleClicksOnOneWayButton = this.handleClicksOnOneWayButton.bind(this);
    this.handleClicksOnReturnButton = this.handleClicksOnReturnButton.bind(this);
    this.handleDepartureDateChange = this.handleDepartureDateChange.bind(this);
    this.handleReturnDateChange = this.handleReturnDateChange.bind(this);
    this.handlePassengerNumberChange = this.handlePassengerNumberChange.bind(this);
    this.handleFromChange = this.handleFromChange.bind(this);
    this.handleDestinationChange = this.handleDestinationChange.bind(this);
    this.handleRangeValueChange = this.handleRangeValueChange.bind(this);
    this.handleClicksOnSearch = this.handleClicksOnSearch.bind(this);
    this.returnFilteredFlights = this.returnFilteredFlights.bind(this);
    this.getFilteredReturnFlights = this.getFilteredReturnFlights.bind(this);
    this.handleClicksOnSearchThisFlight = this.handleClicksOnSearchThisFlight.bind(this);
  }

  componentDidMount () {
    const url = 'http://localhost:3001/flights';
    axios.get(url)
      .then((res) => {
        this.setState({
          data: res.data,
        });
      })
      .catch((err) => {
        console.log('res: ', err);
      });
  }

  handleClicksOnOneWayButton (e) {
    this.setState({
      isReturnFlight: false,
      returnDate: 'not selected',
    });
    indicateActiveButton(e, 'active');
  }

  handleClicksOnReturnButton (e) {
    this.setState({
      isReturnFlight: true,
    });
    indicateActiveButton(e, 'active');
  }

  handleDepartureDateChange (date) {
    this.setState({
      departureDate: formatDate(date),
    });
  }

  handleReturnDateChange (date) {
    this.setState({
      returnDate: formatDate(date),
    });
  }

  handlePassengerNumberChange (numberOfPassengers) {
    this.setState({
      numberOfPassengers,
    });
  }

  handleFromChange (value) {
    this.setState({
      from: value,
    });
  }

  handleDestinationChange (value) {
    this.setState({
      destination: value,
    });
  }

  handleRangeValueChange (rangeValue) {
    if (!this.state.userHasSearched) {
      this.setState({
        rangeValue,
      });
    } else {
      this.setState({
        rangeValue,
        flights: this.returnFilteredFlights(),
      });
    }
  }

  returnFilteredFlights () {
    const filteredFlights = this.state.data.filter((flight) => {
      const details = (flight && flight.details && flight.details[0]);
      if (flight.details.length === 2 || flight.details.length === 3) {
        flight.details.pop();
      }
      return this.validateFlightsData(details, 'from', 'to', this.state.departureDate, flight.price);
    });
    if (this.state.isReturnFlight) {
      this.getFilteredReturnFlights(filteredFlights);
    }
    calculateFlightPrice(filteredFlights);
    return filteredFlights;
  }

  validateFlightsData (details, firstMatchFromOrTo, secondMatchFromOrTo, departure, price) {
    const { from, destination, rangeValue } = this.state;
    const userInputFrom = new RegExp(from, 'i');
    const userInputTo = new RegExp(destination, 'i');
    const userInputDeparture = new RegExp(departure, 'i');
    const flightPrice = Number(price);
    const matchesFrom = userInputFrom.test(details[firstMatchFromOrTo]);
    const matchesTo = userInputTo.test(details[secondMatchFromOrTo]);
    const matchesDeparture = userInputDeparture.test(details.departure);
    const fitsIntoPriceRange = (flightPrice >= rangeValue.min) && (flightPrice <= rangeValue.max);

    return matchesFrom && matchesTo && matchesDeparture && fitsIntoPriceRange;
  }

  getFilteredReturnFlights (filteredFlights) {
    const filteredReturnFlights = flights.filter((flight) => {
      const details = (flight && flight.details && flight.details[0]) || {

      };
      return this.validateFlightsData(details, 'to', 'from', this.state.returnDate, flight.price);
    });

    filteredFlights.forEach((oneWayFlight, i) => {
      if (oneWayFlight.details.length < 2) {
        filteredReturnFlights.forEach((returnFlight, j) => {
          if (oneWayFlight.airlineName === returnFlight.airlineName) {
            filteredFlights[i].details.push(filteredReturnFlights[j].details[0]);
          }
        });
      }
    });
    return filteredFlights;
  }

  handleClicksOnSearch () {
    const { from, destination } = this.state;
    const inputFromIsValid = isValidInput(from);
    const inputToIsValid = isValidInput(destination);

    if (inputFromIsValid && inputToIsValid) {
      this.setState({
        userHasSearched: true,
        flights: this.returnFilteredFlights(),
      });
    }
  }

  handleClicksOnSearchThisFlight (newData) {
    this.setState({
      flights: newData,
    });
  }

  render () {
    const {
      from,
      destination,
      isReturnFlight,
      departureDate,
      returnDate,
      rangeValue,
    } = this.state;

    return (
      <React.Fragment>
        <Header />
        <main>
          <SearchBar
            handleClicksOnOneWayButton={this.handleClicksOnOneWayButton}
            handleClicksOnReturnButton={this.handleClicksOnReturnButton}
            handleDepartureDateChange={this.handleDepartureDateChange}
            handleReturnDateChange={this.handleReturnDateChange}
            isReturnFlight={isReturnFlight}
            departureDate={departureDate}
            returnDate={returnDate}
            onPassengerNumberChange={this.handlePassengerNumberChange}
            onFromChange={this.handleFromChange}
            onDestinationChange={this.handleDestinationChange}
            rangeValue={rangeValue}
            handleRangeValueChange={this.handleRangeValueChange}
            handleClicksOnSearch={this.handleClicksOnSearch}
          />
          <ResultTable
            from={from}
            destination={destination}
            flights={this.state.flights}
            isReturnFlight={isReturnFlight}
            departureDate={departureDate}
            returnDate={returnDate}
            handleClicksOnSearchThisFlight={this.handleClicksOnSearchThisFlight}
          />
        </main>
      </React.Fragment>
    );
  }
}

export default FlightSearchEngineApp;
