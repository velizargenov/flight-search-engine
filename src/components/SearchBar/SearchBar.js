import React from 'react';
import PropTypes from 'prop-types';

import './SearchBar.css';
import Button from '../Button/Button';
import Form from './Form';
import Range from './Range';

const SearchBar = (props) => {
  const {
    handleClicksOnOneWayButton,
    handleClicksOnReturnButton,
    handleDepartureDateChange,
    handleReturnDateChange,
    isReturnFlight,
    departureDate,
    returnDate,
    onPassengerNumberChange,
    onFromChange,
    onDestinationChange,
    handleRangeValueChange,
    handleClicksOnSearch,
    rangeValue,
  } = props;

  return (
    <div className="search-bar">
      <div className="search-bar--buttons">
        <Button
          htmlFor="one-way"
          className="button--search-bar one-way"
          value="One way"
          handleClick={handleClicksOnOneWayButton}
        />
        <Button
          htmlFor="return"
          className="button--search-bar return active"
          value="Return"
          handleClick={handleClicksOnReturnButton}
        />
      </div>
      <Form
        handleDepartureDateChange={handleDepartureDateChange}
        handleReturnDateChange={handleReturnDateChange}
        isReturnFlight={isReturnFlight}
        departureDate={departureDate}
        returnDate={returnDate}
        onPassengerNumberChange={onPassengerNumberChange}
        onFromChange={onFromChange}
        onDestinationChange={onDestinationChange}
        handleClicksOnSearch={handleClicksOnSearch}
      />
      <Range
        rangeValue={rangeValue}
        handleRangeValueChange={handleRangeValueChange}
      />
    </div>
  );
};

SearchBar.defaultProps = {
  handleClicksOnOneWayButton: () => {},
  handleClicksOnReturnButton: () => {},
  handleDepartureDateChange: () => {},
  handleReturnDateChange: () => {},
  isReturnFlight: true,
  departureDate: '18-01-2018',
  returnDate: '20-01-2018',
  onPassengerNumberChange: () => {},
  onFromChange: () => {},
  onDestinationChange: () => {},
  handleRangeValueChange: () => {},
  handleClicksOnSearch: () => {},
  rangeValue: {
    max: 200,
    min: 0,
  },
};

SearchBar.propTypes = {
  handleClicksOnOneWayButton: PropTypes.func,
  handleClicksOnReturnButton: PropTypes.func,
  handleDepartureDateChange: PropTypes.func,
  handleReturnDateChange: PropTypes.func,
  isReturnFlight: PropTypes.bool,
  departureDate: PropTypes.string,
  returnDate: PropTypes.string,
  onPassengerNumberChange: PropTypes.PropTypes.func,
  onFromChange: PropTypes.PropTypes.func,
  onDestinationChange: PropTypes.PropTypes.func,
  handleRangeValueChange: PropTypes.PropTypes.func,
  handleClicksOnSearch: PropTypes.PropTypes.func,
  rangeValue: PropTypes.shape({
    max: PropTypes.number,
    min: PropTypes.number,
  }),
};

export default SearchBar;
