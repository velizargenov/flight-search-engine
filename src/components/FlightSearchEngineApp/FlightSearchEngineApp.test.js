import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import FlightSearchEngineApp from './FlightSearchEngineApp';
import Header from '../Header/Header';
import SearchBar from '../SearchBar/SearchBar';
import ResultTable from '../ResultTable/ResultTable';

let wrapper;
beforeEach(() => {
  wrapper = shallow(<FlightSearchEngineApp />);
});

describe('FlightSearchEngineApp Component', () => {
  describe('Rendering', () => {
    it('matches snapshot', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render without crashing', () => {
      expect(wrapper.length).toBe(1);
    });
    it('should render Header, SearchBar and ResultTable components', () => {
      expect(wrapper.containsAllMatchingElements([
        <Header />,
        <SearchBar />,
        <ResultTable />,
      ]));
    });
    it('should render one of each Header, SearchBar, ResultTable components', () => {
      expect(wrapper.find(Header).length).toBe(1);
      expect(wrapper.find(SearchBar).length).toBe(1);
      expect(wrapper.find(ResultTable).length).toBe(1);
    });
    it('should render one <React.Fragment> element', () => {
      expect(wrapper.find('main').length).toBe(1);
    });
    it('should render one <main> element', () => {
      expect(wrapper.find('main').length).toBe(1);
    });
  });
});
