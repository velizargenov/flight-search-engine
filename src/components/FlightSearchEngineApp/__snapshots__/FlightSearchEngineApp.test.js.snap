// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightSearchEngineApp Component Rendering matches snapshot 1`] = `
<React.Fragment>
  <Header />
  <main>
    <SearchBar
      departureDate="18-01-2018"
      handleClicksOnOneWayButton={[Function]}
      handleClicksOnReturnButton={[Function]}
      handleClicksOnSearch={[Function]}
      handleDepartureDateChange={[Function]}
      handleRangeValueChange={[Function]}
      handleReturnDateChange={[Function]}
      isReturnFlight={true}
      onDestinationChange={[Function]}
      onFromChange={[Function]}
      onPassengerNumberChange={[Function]}
      rangeValue={
        Object {
          "max": 200,
          "min": 0,
        }
      }
      returnDate="not selected"
    />
    <ResultTable
      departureDate="18-01-2018"
      destination=""
      flights={Array []}
      from=""
      isReturnFlight={true}
      returnDate="not selected"
    />
  </main>
</React.Fragment>
`;
